### PR TITLE
[EC-16 / EC-86] Implement Groups Table Endpoints

### PR DESCRIPTION
**Pull request by shane-melton**
## Type of change

```
- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective
Updates the Server to support showing Groups **and** their associated collections in the same table on the Groups tab in the Org admin console. It also adds support for doing bulk deletion operations on Groups. ([Client PR here](https://github.com/bitwarden/clients/pull/3563)) Also includes modifying a groups users during `POST` or `PUT` requests.

## Code changes

* **src/Api/Controllers/GroupsController.cs:** 
  * Include Collection details in list groups endpoint
  * Add new bulk delete endpoint
  * Use the new `DeleteGroupCommand`
  * Include list of user Ids in `PUT` and `POST` group endpoints

* **src/Core/OrganizationFeatures/OrganizationGroups/DeleteGroupCommand.cs:** 
  * Add command to bulk delete groups
  * Also refactor delete single group functionality out of GroupService 
 
* **src/Api/Models/Request/GroupRequestModel.cs:**
  * Add optional `Users` property to the request model for configuring group members during group creation/modification
  * Add new bulk request model for deleting multiple groups

* **src/Core/Services/Implementations/GroupService.cs:**
  * Update `SaveAsync()` to support **optionally** modifying a group's users during creation/modification 

* **src/Infrastructure.Dapper/Repositories/GroupRepository.cs:**
  * Add get many groups by many ids method
  * Add get many groups with collection details method
  * Add delete many groups method

* **src/Core/OrganizationFeatures/OrganizationServiceCollectionExtensions.cs:** Register the new `DeleteGroupCommand`

* **Various Stored Procedures:** Add the necessary stored procedures for the new `GroupRepository` methods

* **src/Sql/dbo/Stored Procedures/User_BumpAccountRevisionDateByOrganizationIds.sql:** Extended version of the `User_BumpAccountRevisionDateByOrganizationId` sproc that accepts multiple organization ids as input.

* **util/Migrator/DbScripts/2022-09-15_00_ReadGroupsWithCollectionsByOrgId.sql:** Migration script to add the new stored procedures

* **bitwarden_license/src/Scim/Controllers/v2/GroupsController.cs:** Use the new `DeleteGroupCommand`

* **src/Api/Public/Controllers/GroupsController.cs:** Include collection details in public Groups API endpoint (as mentioned in an outstanding TODO)

## Before you submit

- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- If making database changes - make sure you also update Entity Framework queries and/or migrations
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

File: src/Core/OrganizationFeatures/OrganizationGroups/DeleteGroupCommand.cs
Comment (vincentsalucci) [inline src/Core/OrganizationFeatures/OrganizationGroups/DeleteGroupCommand.cs:29]: Thoughts on moving this logic out of the command so it only needs to take IDs and perform the action/return the result? Maybe into a helper or even in the controller.

cc: @MGibson1 Does this suggestion make sense per our conversation?

Comment (shane-melton) [reply src/Core/OrganizationFeatures/OrganizationGroups/DeleteGroupCommand.cs:29]: My initial feeling is that this logic belongs in the command as it should likely be applied every time multiple groups are deleted (always ensuring the deleted groups belong to the expected organization). Otherwise, it will just be repeated in each calling controller/service which provides more surface area for mistakes, imo.

With that said, I'm open to hearing more on why you think it should be moved out of the command, I could be missing some context related to how we want CQRS commands to work.

Comment (MGibson1) [reply src/Core/OrganizationFeatures/OrganizationGroups/DeleteGroupCommand.cs:29]: I'd like to see grabbing of organization and groups outside of the command, then a guard that throws if a group doesn't belong to the given org is reasonable.

However, even more reasonable would be not requiring an org at all. The controller should validate permissions to delete any groups it gives in, then at the command level it shouldn't care at all about the orgs.

It seems the way the sproc is written right now limits deleteMany to a single organization. I don't really understand why In a repository sense, there's no reason to restrict deletes of a group to a single org at a time. this is a business logic concern at the controller level (if even there, as long as you have permission, I don't really care if you delete for a single org or many)

File: src/Infrastructure.EntityFramework/Repositories/GroupRepository.cs
Comment (eliykat) [inline src/Infrastructure.EntityFramework/Repositories/GroupRepository.cs:121]: Why does this use `ToArray` instead of `ToList`, which is what we seem to use elsewhere here? (and is the more common choice afaik)

File: src/Infrastructure.EntityFramework/Repositories/GroupRepository.cs
Comment (eliykat) [inline src/Infrastructure.EntityFramework/Repositories/GroupRepository.cs:107]: I don't think this is grouping the collections by group, it's just associating every collection with every group.

**Review by eliykat — CHANGES_REQUESTED** [review]
This generally looks pretty good, I've left some feedback below. It's been a while since I've written DB queries, please get a second review on those (i.e. repository and sproc changes only).

File: src/Core/Repositories/IGroupRepository.cs
Comment (eliykat) [reply src/Core/Repositories/IGroupRepository.cs:20]: Not a hold up for this PR. Just wondering.

File: src/Api/Controllers/GroupsController.cs
Comment (shane-melton) [reply src/Api/Controllers/GroupsController.cs:144]: We previously only had an endpoint to delete a single group at a time. This is adding the bulk delete endpoint accepting a list of group Ids in the request body (instead of a single group Id in the url)

File: src/Api/Controllers/GroupsController.cs
Comment (shane-melton) [reply src/Api/Controllers/GroupsController.cs:160]: The front-end doesn't require all the details be returned, but I was trying to keep it consistent with the other admin service endpoints which return all the details during bulk deletion (`OrganizaitonUsersController` and `ProviderUsersController`). I didn't see any that just return `id`s, but I can do that if that's what we want to practice going forward.

I also see the `CiphersController` doesn't return anything for any of the delete many operations.

File: src/Api/Controllers/GroupsController.cs
Comment (shane-melton) [reply src/Api/Controllers/GroupsController.cs:59]: I talked with @vincentsalucci a bit about this earlier, and we didn't see any concern for backwards compatibility as this change only added a new property `Collections` to the response.

File: src/Core/OrganizationFeatures/OrganizationGroups/Interfaces/IDeleteGroupCommand.cs
Comment (shane-melton) [reply src/Core/OrganizationFeatures/OrganizationGroups/Interfaces/IDeleteGroupCommand.cs:3]: Good point, I'll get that renamed

File: src/Core/OrganizationFeatures/OrganizationGroups/DeleteGroupCommand.cs
Comment (shane-melton) [reply src/Core/OrganizationFeatures/OrganizationGroups/DeleteGroupCommand.cs:26]: We'd probably be fine not converting to an array, but an `IEnumerable` does not guarantee it supports [multiple enumeration](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1851), so the conversion to an Array guarantees the `IEnumerable` is only enumerated a single time in the method. 

We could also change the parameter type to something like `IList` or `ICollection` if we wanted to avoid the need for a conversion, but it seemed like we typically pass `IEnumerable`s around.

File: src/Sql/dbo/Stored Procedures/Group_DeleteByIds.sql
Comment (shane-melton) [reply src/Sql/dbo/Stored Procedures/Group_DeleteByIds.sql:34]: The _single_ `Group_DeleteById` sproc bumped all members belonging to the organization so I transferred that logic over to the batch sproc here. If that's not the desired effect I can look into updating both to only bump the revision date for group members. (I'm just not familiar with all the processes that depend on revision date being bumped to know for sure myself at the moment)

File: src/Core/Services/Implementations/GroupService.cs
Comment (shane-melton) [reply src/Core/Services/Implementations/GroupService.cs:32]: Totally agree here, but I think its best to create a separate task for it at this time.

File: src/Infrastructure.EntityFramework/Repositories/GroupRepository.cs
Comment (shane-melton) [reply src/Infrastructure.EntityFramework/Repositories/GroupRepository.cs]: Whoops, good catch! Fixed now.

File: src/Api/Controllers/GroupsController.cs
Comment (eliykat) [reply src/Api/Controllers/GroupsController.cs:59]: good point!

File: src/Sql/dbo/Stored Procedures/User_BumpAccountRevisionDateByOrganizationIds.sql
Comment (justindbaur) [inline src/Sql/dbo/Stored Procedures/User_BumpAccountRevisionDateByOrganizationIds.sql:10]: I'm curious if there is another example of updating many accounts at once if we call `GETUTCDATE` once and use that value for them all or we call it live for each row. If there is an example of that and we only call it once then I'd like to switch this one to that but if there is no example or this is what the example does then this is fine.

**Review by justindbaur — APPROVED** [review]
SPROCS looks good to me minus one question, and I will leave if the question results in a change up to you.

File: src/Sql/dbo/Stored Procedures/User_BumpAccountRevisionDateByOrganizationIds.sql
Comment (shane-melton) [reply src/Sql/dbo/Stored Procedures/User_BumpAccountRevisionDateByOrganizationIds.sql:10]: I did look into that, and saw we did it live for each row in both `User_BumpAccountRevisionDateByProviderUserIds` and `User_BumpAccountRevisionDateByOrganizationUserIds` so I kept with the current practice.

File: src/Core/OrganizationFeatures/OrganizationGroups/DeleteGroupCommand.cs
Comment (eliykat) [reply src/Core/OrganizationFeatures/OrganizationGroups/DeleteGroupCommand.cs]: That makes sense, thanks! It may not be necessary once you make a bulk `LogGroupEventsAsync` method though, you'll only be enumerating it once in this method for the `Select` method.

File: src/Api/Controllers/GroupsController.cs
Comment (eliykat) [reply src/Api/Controllers/GroupsController.cs:160]: The bulk *users operations return individualised success or error message for each user. (Try to bulk confirm users on your local instance and you'll see what I mean.) And we need to do that because we allow each operation on a user to succeed or fail separately, we don't abort a whole operation because some users couldn't be updated.

Let's only send information back to the client if we need it. So:
* if the UI allows for individualised error messages like this, or if there's a chance that only some groups will be affected but not all, then send a response back to the client
* if not, if it'll either succeed or fail and we can always know from the client what happened, then I think we just return an empty response (or 200 OK, whatever)

Comment (shane-melton) [reply src/Api/Controllers/GroupsController.cs:160]: Sounds reasonable to me! The delete endpoint/command is all or nothing, so there shouldn't be a need to return anything to the client besides a 200 OK.

File: src/Infrastructure.EntityFramework/Repositories/GroupRepository.cs
Comment (eliykat) [reply src/Infrastructure.EntityFramework/Repositories/GroupRepository.cs]: I was just about to ask you to change this to a list, but you're way ahead of me ;)

File: src/Sql/dbo/Stored Procedures/Group_ReadByIds.sql
Comment (eliykat) [reply src/Sql/dbo/Stored Procedures/Group_ReadByIds.sql:10]: resolving as Justin has kindly double checked this

File: src/Sql/dbo/Stored Procedures/Group_DeleteByIds.sql
Comment (eliykat) [reply src/Sql/dbo/Stored Procedures/Group_DeleteByIds.sql:34]: Ok fair enough, let's stick with the existing pattern.

File: src/Core/Repositories/IGroupRepository.cs
Comment (eliykat) [reply src/Core/Repositories/IGroupRepository.cs:20]: I'm thinking this is fine, it's the nature of server that it's often getting Guids only and needs to be able to deal with those directly.

File: src/Core/Services/Implementations/GroupService.cs
Comment (eliykat) [reply src/Core/Services/Implementations/GroupService.cs:32]: Please link a Jira ticket so I can see it's tracked for follow-up :)

File: src/Api/Controllers/GroupsController.cs
Comment (eliykat) [reply src/Api/Controllers/GroupsController.cs:73]: Looking at this again today, I'm less concerned about it. Your call whether you think it's worth doing.

**Review by eliykat — COMMENTED** [review]
Reviewed latest changes and responded to any comments. Marked everything as resolved that has been resolved, just a few things left outstanding from the original review.

File: src/Api/Controllers/GroupsController.cs
Comment (shane-melton) [reply src/Api/Controllers/GroupsController.cs]: Done.

File: src/Api/Controllers/GroupsController.cs
Comment (shane-melton) [reply src/Api/Controllers/GroupsController.cs:73]: I agree that tuples aren't always ideal due to the use of generic names like `Item1` and `Item2`, but I don't currently see a benefit for moving the tuple deconstruction into the `GroupDetailsResponseModel` at this time

File: src/Sql/dbo/Stored Procedures/User_BumpAccountRevisionDateByOrganizationIds.sql
Comment (shane-melton) [reply src/Sql/dbo/Stored Procedures/User_BumpAccountRevisionDateByOrganizationIds.sql:10]: Gonna mark this as resolved per
> [if] this is what the example does then this is fine.

File: src/Core/OrganizationFeatures/OrganizationGroups/DeleteGroupCommand.cs
Comment (shane-melton) [reply src/Core/OrganizationFeatures/OrganizationGroups/DeleteGroupCommand.cs]: Unfortunately, it's still "required" as all methods (e.g. `LogGroupEventsAsync()`) that accept `IEnumerable` are, [by default](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1851#configure-the-default-assumption-of-methods-take-ienumerable-parameters), assumed to enumerate the argument.

But it is much cleaner now without the `for` loop and the multiple `awaits`

File: src/Core/OrganizationFeatures/OrganizationGroups/DeleteGroupCommand.cs
Comment (shane-melton) [reply src/Core/OrganizationFeatures/OrganizationGroups/DeleteGroupCommand.cs]: Done, good call!

File: src/Core/Services/Implementations/GroupService.cs
Comment (shane-melton) [reply src/Core/Services/Implementations/GroupService.cs]: Completed with abb98ee3afe99731ef5fb7cb926d4738195f1f2a

File: src/Core/OrganizationFeatures/Groups/DeleteGroupCommand.cs
Comment (eliykat) [inline src/Core/OrganizationFeatures/Groups/DeleteGroupCommand.cs:26]: I'm still a bit hung up on this. I think I understand the general concern: if `groups` is an `IEnumerable` that defers enumeration and does not cache enumeration results, then this will read from the database each time we enumerate it in the method (i.e. in the 2 `Select` statements below).

At the moment, we generally pass around `IEnumerables` without regard for this risk, so is this a blind spot in our practices? Should we just force enumeration in repository methods to be safe? I don't know, I think this is worth raising with the team.

For now, in this PR, I think this should be enforced by method parameter types - it's basically a type system concern. If the method requires a list/array/whatever then it should state that in its method signature.

Our repository methods return `ICollection`, is this subject to deferred execution? I'm trying to think of the most convenient/general parameter type to require. What do you think?

**Review by eliykat — CHANGES_REQUESTED** [review]

File: src/Core/Services/Implementations/GroupService.cs
Comment (shane-melton) [reply src/Core/Services/Implementations/GroupService.cs]: Renamed to `usersToAddToGroup` to be more explicit about its purpose

File: src/Core/Services/Implementations/GroupService.cs
Comment (shane-melton) [reply src/Core/Services/Implementations/GroupService.cs:89]: We can't really move this down, as it must come before `await _groupRepository.UpdateUsersAsync(group.Id, newUserIds);` to ensure we get the list of original userIds before we update them.

File: src/Core/OrganizationFeatures/Groups/DeleteGroupCommand.cs
Comment (shane-melton) [reply src/Core/OrganizationFeatures/Groups/DeleteGroupCommand.cs]: > At the moment, we generally pass around `IEnumerables` without regard for this risk, so is this a blind spot in our practices?

I believe this might be a blind spot, however, I don't believe we typically use deferred execution, so it's likely not a significant concern. Should we start using deferred execution more frequently, I think it would be beneficial to consider / keep in mind going forward.

> Our repository methods return `ICollection`, is this subject to deferred execution? I'm trying to think of the most convenient/general parameter type to require. What do you think?

Yep, `ICollection` makes the most sense to me as well, and is not subject to deferred execution. I went ahead and switched the `DeleteManyAsync` signature to use `ICollection<Group>` and removed the `.toArray()`  in the method body.

File: src/Api/Public/Controllers/GroupsController.cs
Comment (shane-melton) [reply src/Api/Public/Controllers/GroupsController.cs:83]: Yep, no special requirements for **non-breaking** changes, so we should be all clear

File: src/Core/Services/Implementations/GroupService.cs
Comment (eliykat) [reply src/Core/Services/Implementations/GroupService.cs:89]: I was just referring to `originalUserIds` here, which isn't reference in the group repository call. But it's also not a big deal so I won't worry about it.

**Review by eliykat — APPROVED** [review]

**Pull request merged**
